### ISSUE TITLE
GetLiteralTokens slow with large command texts

### ISSUE THREAD

**Issue by NinoFloris** [StackExchange/Dapper#1484]
For a rarely used feature the overhead is really too big -- making up 7% of our response times -- and it's all spent in the regex engine.

<img width="1021" alt="GetLiteralTokens bench" src="https://user-images.githubusercontent.com/4218809/84599441-627bb280-ae72-11ea-9f9f-0126d3e869d4.png">

Could this be placed behind a flag or in a `DynamicParameters` derived type?

If so I'm willing to PR it.

**Comment by NickCraver** [comment]
We could probably check for the existence of `{=` in the string at all first as a cheap check for the common case there. Are you using _no_ literals? The literal isn't based on the parameter really, it's from the query side so we need to base it on that as well...can't be a property on the params (plus that would make it much harder to consume from an API standpoint).

Overall I'd propose: don't start with changing the API. Start with optimizing the current API which would go a very long way.

**Comment by NinoFloris** [comment]
I'm not sure why the feature exists at all to be honest 😀.

If need differing literals/trusted sql fragments in my command text I'd use string interpolation before passing the final form to Dapper, and we do this everywhere. 

This is really a hidden (nobody outside SO probably knows it exists) avenue for sql injection, which when needed should be brought front and center during command text construction IMO.

Also, interpolating and building the string once is also more efficient.

Now that's said though, adding a check for `{=` would indeed help, Contains is nicely vectorized too.

**Comment by NickCraver** [comment]
I agree `{=` check is the path forward here, but I wanted to correct some assumptions stated above to get on the same page:

> I'm not sure why the feature exists at all to be honest 😀.

For things like filtered indexes, which a parameter can't match in a query plan...while also using literals and enums for easy code usage finding.

> If we need differing literals/trusted sql fragments in command text I'd use string interpolation before passing the final form to Dapper, and we do this everywhere.

That's a good way to get SQL injection. Dapper doesn't allow that to happen, e.g. it doesn't work on string types. It's a safety mechanism.

> This is really a hidden (nobody outside SO probably knows it exists) avenue for sql injection, which when needed should be brought front and center during command text construction IMO.

No it's not, see above - we explicitly don't allow that case.

> Also, interpolating and building the string once is also more efficient.

Sure, it that works for you, but most people would just interpolate. Note that the EF Core project took an `IFormattableString` overload to support parameterization then back away in the next major version, because it's ripe for SQL injection. I made them some samples for that discussion here: https://github.com/NickCraver/EFCoreInjectionSample.

**Comment by NinoFloris** [comment]
Alright.

Let me also correct some things as it seems I've made myself misunderstood.

>That's a good way to get SQL injection

How's that? Sometimes you need, say, an additional join in your SQL, user only controls inclusion or not. The fragments themselves are entirely static, hence trusted like the rest of the sql. If there are not too many combinations we usually cache this on a static, but sometimes this is on demand.

>Sure, it that works for you, but most people would just interpolate

Isn't that what I said?

We have parameters for a reason of course, I'm not sure how I've found myself carrying the opinion that I'm advocating anything but parameterization?

**Comment by NickCraver** [comment]
> How's that?

You seem to based this on all *your* strings being added together as static and trusted. That's both unassured and very easy to miss in a code review. For example: we do now allow interpolation for SQL strings without an explicit exception where it's the best path. Even then, we opt for an explicit `string.Format` so it's clearer. We even have a static code analyzer enforcing this every build.

> Isn't that what I said?

Yes, it is. Note that I meant it as a negative. Defaulting to this behavior is an easy way to get a string from input injected, especially for a newer dev who doesn't know any better.

> This is a very SQL Server centric observation btw

Somewhat, but those specialized plans don't do what most people think. And with filtered indexes, that's not the case unless something changed in the latest major version.

Overall the argument seems to be based on safety of interpolation is with **static** strings, at least as pitched. That's an unsafe assumption. Especially in a code review context where we see one line and don't see where the string above came from. *Or* we could be safe and not allow or enable it at all. That's what we do. You can take whatever practices you deem safe in your app, and best of luck with doing so - but in a library context we have to strive, if we wish to be responsible maintainers, of defaulting users to the pit of success.

I won't even dive into how one might replace your string format or overload rules or formatters for global query injection with any formatter as an attack surface. Let's just please agree: there are issue with this. If you are willing to accept those, then alrighty. Dapper isn't though.

**Comment by NickCraver** [comment]
Note: if you'd like to read further here, I suggest perusing #838 for why we don't enable it like EF Core did for a while :)

**Comment by NinoFloris** [comment]
>those specialized plans don't do what most people think

Interesting, can you tell me more or point me to a resource?

>That's an unsafe assumption

Not sure how these static strings become any more or less unsafe when interpolating vs a string format, but I'm fine seeing string.Format too.

To clarify everything I've said was really only in the context of static sql fragments, not parameter interpolation like the EF example, which I followed closely. 

I'm also not saying literals like the feature enables are useless from a security perspective but this really hinges on whether you need them at all. In our case we never ever inject user input - which literals are apparently used for - there's just no need for it with PG. 

When we really limit the context of my arguments to static strings, doing string.Format or interpolation into a string local or field both have the same outcome, one over the other can be argumented for or against but this is ultimately 'taste'.
